Enforce use of throwing `AbsolutePath` initializer

neonichu opened the pull request:
In support of https://github.com/apple/swift-package-manager/pull/5797

neonichu commented on Sources/TSCTestSupport/FileSystemExtensions.swift:
```diff
@@ -72,3 +72,19 @@ extension FileSystem {
 }
+
+public extension AbsolutePath {
+    init(path: StaticString) {
```
Without an argument label, the `String` version of the initializer seemed to always get chosen, so I ended up just adding an argument label. These initializers are a convenience for tests which often deal with static paths.

neonichu commented:
@swift-ci please test

tomerd commented on Sources/TSCUtility/PkgConfig.swift:
```diff
@@ -47,12 +47,14 @@ public struct PCFileFinder {
-        AbsolutePath("/usr/share/pkgconfig"),
-    ]
+    static func searchPaths() throws -> [AbsolutePath] {
+        return try [
```
does this not use the non-throwing static init?

tomerd commented on Sources/TSCUtility/Platform.swift:
```diff
@@ -49,19 +49,21 @@ public enum Platform: Equatable {
-            return .linux(.fedora)
-        }
+        do {
+            if try fileSystem.isFile(AbsolutePath("/etc/debian_version")) {
```
should we have a non-throwing initializer for static string?

neonichu replied:
I'd like to keep that to tests since it is pretty rare for us to construct paths from static strings.

tomerd reviewed (approved):
looks great. couple of small questions

neonichu commented:
```
/home/build-user/swift-tools-support-core/Sources/TSCBasic/FileSystem.swift:358:33: error: extraneous argument label 'validating:' in call
        return try? AbsolutePath(validating: String(cString: fsr))
                                ^~~~~~~~~~~~~
```

Interesting, not sure why this didn't show up locally

neonichu commented:
@swift-ci please test

neonichu commented:
@swift-ci please test

neonichu commented:
@swift-ci please test

neonichu commented:
Looks like at least on macOS we're reaching the swift-driver build now.

neonichu commented:
@swift-ci please test

neonichu commented:
Updated the approach slightly here, I am keeping `AbsolutePath(validating:)` and changed the relative path one to `AbsolutePath(validating:relativeTo:)`. These changes, plus an extension that offers the non-throwing variants as deprecated, allow clients to move to the throwing variants piecemeal. 

In a future PR, we can remove the deprecated initializers and `validating` argument labels.

neonichu commented:
@swift-ci please test

neonichu commented:
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/swift-tools-support-core/pull/353
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please test

neonichu commented:
https://github.com/apple/swift-package-manager/pull/5797
https://github.com/apple/swift-driver/pull/1204
https://github.com/apple/sourcekit-lsp/pull/655
@swift-ci please test

WowbaggersLiquidLunch commented on Sources/TSCBasic/FileSystem.swift:
```diff
@@ -353,7 +355,7 @@ private class LocalFileSystem: FileSystem {
         let fsr: UnsafePointer<Int8> = cwdStr.fileSystemRepresentation
         defer { fsr.deallocate() }
 
-        return try? AbsolutePath(validating: String(cString: fsr))
```
It seems to me this broke SwiftPM.
